lib/puppet/transaction.rb

```diff
@@ -361,7 +361,7 @@ def resources_by_provider(type_name, provider_name)
       @catalog.vertices.each do |resource|
         if resource.class.attrclass(:provider)
           prov = resource.provider && resource.provider.class.name
-          @resources_by_provider[resource.type][prov][resource.name] = resource
+          @resources_by_provider[resource.type][prov][resource.title] = resource
```

Comment:
Since resource title doesn't always equal name, I'm pretty sure this will be break something.

Reply:
Looks like puppet originally prefetched using `title` (see commit c35d07b4183)

```ruby
                prefetchers[provider.class][resource.title] = resource
```
Later there was a bug reported in the mount provider [redmine #2013](https://web.archive.org/web/20140325080254/https://projects.puppetlabs.com/issues/2013) And puppet was changed to use `name` in 9577d3af929727ac1e7b5e7e54e4491990d55995 as "ParsedFile types seem to be the main one that suffers from (mismatch between type and title)"

```diff
-                prefetchers[provider.class][resource.title] = resource
+                prefetchers[provider.class][resource.name] = resource
```

I'm not convinced that changing the transaction to use `name` was the right fix (as opposed to a bug in the mount provider), but that ship sailed awhile ago. We'll need a different way for parsed file providers to opt-into the new behavior.